Item 10253: Remove experimental feature to officially switch to new Grid Filtering UX changes

cnathe opened the pull request:
#### Rationale
The app GridPanel OmniBox has served us very well for many reason, but it is time for a change. The related PRs remove the experimental feature flag to switch over to the new grid filtering UX (filter modal from sample finder, grid column header menu options for add/remove filter, grid filter status display in header, search input in grid button bar, etc.). With that change, this PR updates the test components/pages so that the filtering for an app grid is done via the GridFilterModal instead of the OmniBox (along with any grid search and sorts).

#### Related Pull Requests
* https://github.com/LabKey/labkey-ui-components/pull/804
* https://github.com/LabKey/testAutomation/pull/1076
* https://github.com/LabKey/platform/pull/3247
* https://github.com/LabKey/biologics/pull/1245
* https://github.com/LabKey/inventory/pull/418
* https://github.com/LabKey/sampleManagement/pull/918
* https://github.com/LabKey/provenance/pull/103

#### Changes
* Remove references to QueryGridPanel in test locators
* update css selectors in FilterFacetedPanel
* test helper Grid updates for removal of OmniBox in favor of FilterStatus display
* Factor out GridFilterModal and change EntityFieldFilterModal to extend it

cnathe committed:
Remove references to QueryGridPanel in test locators

cnathe committed:
Merge branch 'develop' into fb_gridFilterUXEnabled

cnathe committed:
update css selectors in FilterFacetedPanel

cnathe committed:
Grid updates for removal of OmniBox in favor of FilterStatus display
- search moved to grid bar search input
- sorts and filters moved to using the column header menus

cnathe committed:
Temp change to EntityFieldFilterModal to support Sample Finder and Grid Panel cases

cnathe committed:
Merge branch 'develop' into fb_gridFilterUXEnabled

cnathe committed:
fix after merge

cnathe committed:
Factor out GridFilterModal and change EntityFieldFilterModal to extend it

cnathe committed:
omnibox-value => filter-status-value

cnathe committed:
GridFilterModal confirm expecting error

cnathe committed:
grid filter with multiple expressions for a given column

labkey-tchad commented on src/org/labkey/test/components/ui/FilterStatus.java:
```diff
@@ -0,0 +1,79 @@
+
+import java.util.List;
+
+public class FilterStatus extends WebDriverComponent<FilterStatus.ElementCache>
```
I don't think this needs to be a component. It can be a plain `WebElement` in the QueryGrid element cache.
```
final WebElement filterStatusPanel = Locator.css("div.grid-panel__filter-status").findWhenNeeded(this);
```
Then move the `FilterStatusValue` getters into that element cache as well.

cnathe replied:
Done. This was originally the `OmniBox` test component that I renamed to `FilterStatus`. You are right, after removing several things, there isn't much left.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/GridFilterModal.java:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    @Override
+    protected void waitForReady(ModalDialog.ElementCache ec)
+    {
+        getWrapper().shortWait().until(ExpectedConditions.visibilityOf(listItem.findElementOrNull(elementCache().fieldsSelectionPanel)));
+    }
```
`visibilityOf` will NPE if `listItem.findElementOrNull` returns null. `findWhenNeeded` is what you want; it will play nicely with `visibilityOf` when the element isn't present.
```suggestion
    protected void waitForReady(ModalDialog.ElementCache ec)
    {
        getWrapper().shortWait().until(ExpectedConditions.visibilityOf(listItem.findWhenNeeded(elementCache().fieldsSelectionPanel)));
    }
```

cnathe replied:
Done

labkey-tchad commented on src/org/labkey/test/components/ui/grids/GridFilterModal.java:
```diff
@@ -0,0 +1,139 @@
+        return elementCache().filterFacetedPanel;
+    }
+
+    /**
+     * Save current changes to the search criteria.
+     * Throw <code>IllegalStateException</code> if the save button is disabled because no changes have been made.
+     */
+    public void confirm()
+    {
+        confirm(false);
+    }
+
+    public void confirm(boolean errorExpected)
+    {
+        if (!elementCache().submitButton.isEnabled())
+        {
+            throw new IllegalStateException("Confirmation button is not enabled.");
+        }
+
+        if (errorExpected)
+        {
+            elementCache().submitButton.click();
+            getWrapper().shortWait().until(ExpectedConditions.visibilityOf(elementCache().errorAlert));
+        }
+        else
+        {
+            _linkedComponent.doAndWaitForUpdate(() -> {
+                elementCache().submitButton.click();
+                waitForClose();
+            });
+        }
+    }
```
The preferred pattern is to have separate methods for distinct behavior instead of using a flag. In this case it would allow the confirm with error to return the error.
```suggestion
    /**
     * Save current changes to the search criteria.
     * Throw <code>IllegalStateException</code> if the save button is disabled because no changes have been made.
     */
    public void confirm()
    {
        _linkedComponent.doAndWaitForUpdate(() -> {
            clickConfirm();
            waitForClose();
        });
    }

    public String confirmExpectingError()
    {
        clickConfirm();
        return getWrapper().shortWait().until(ExpectedConditions.visibilityOf(elementCache().errorAlert)).getText();
    }

    protected void clickConfirm()
    {
        if (!elementCache().submitButton.isEnabled())
        {
            throw new IllegalStateException("Confirmation button is not enabled.");
        }

        elementCache().submitButton.click();
    }
```

cnathe replied:
Sounds good. Thanks.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/QueryGrid.java:
```diff
@@ -185,43 +188,45 @@ public void doAndWaitForUpdate(Runnable func)
 
     /**
-     * clears search, sort, and filter expressions via the omnibox
+     * clears grid filter expressions added by the user (i.e. from the filter modal)
      */
-    public QueryGrid clearSortsAndFilters()
+    public QueryGrid clearUserDefinedFilters()
```
Why the distinction of "user defined". What other filters might there be?
I would prefer the brevity of just `clearFilters` if possible. And just have the JavaDoc make any necessary clarifications.

cnathe replied:
The grid could be filtered by the context (i.e. baseFilters applied via code directly like in the "view samples for a given source" grid) and the grid can be filtered by the saved view. Those two cannot be cleared/removed, so I renamed this method accordingly. This likely isn't meaningful from the test side, so I'll rename to `clearFilters`.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/QueryGrid.java:
```diff
@@ -185,43 +188,45 @@ public void doAndWaitForUpdate(Runnable func)
      */
-    public QueryGrid filterOn(String columnName, OmniBox.FilterOperator operator, String value)
+    public QueryGrid sortOn(String column, SortDirection direction)
     {
-        getOmniBox().setFilter(columnName, operator, value);
+        sortColumn(column, direction);
         return this;
     }
```
Can we get rid of `sortOn` so that `QueryGrid` doesn't have redundant sorting methods?

cnathe replied:
Yes. I have removed `sortOn` and changed callers to just call `sortColumn` directly.

labkey-tchad commented on src/org/labkey/test/components/ui/grids/ResponsiveGrid.java:
```diff
@@ -107,9 +109,78 @@ public T sortColumnDescending(String columnLabel)
+    private GridFilterModal initFilterColumn(String columnLabel, Filter.Operator operator, Object value)
+    {
+        clickColumnMenuItem(columnLabel, "Filter...", false);
+        GridFilterModal filterModal = new GridFilterModal(getDriver(), getThis()::doAndWaitForUpdate);
```
ResponsiveGrid implements `UpdatingComponent` interface that the constructor takes. This shouldn't need to use a method reference.
```suggestion
        GridFilterModal filterModal = new GridFilterModal(getDriver(), this);
```

cnathe replied:
Yup, good catch. Done

labkey-tchad commented on src/org/labkey/test/components/ui/search/EntityFieldFilterModal.java:
```diff
@@ -4,32 +4,25 @@
-        super.waitForReady(ec);
-        getWrapper().shortWait().until(ExpectedConditions
-                .visibilityOf(listItem.findWhenNeeded(elementCache().querySelectionPanel)));
+        getWrapper().shortWait().until(ExpectedConditions.visibilityOf(listItem.findElementOrNull(elementCache().querySelectionPanel)));
```
Need to use `findWhenNeeded` with `visibilityOf`
```suggestion
        getWrapper().shortWait().until(ExpectedConditions.visibilityOf(listItem.findWhenNeeded(elementCache().querySelectionPanel)));
```

cnathe replied:
Done

labkey-tchad commented on src/org/labkey/test/tests/component/GridPanelTest.java:
```diff
@@ -221,86 +221,51 @@ public void testOmniboxFilterInMultiPageSelectionSet() throws Exception
         TestDataGenerator sampleSetDataGenerator = SampleTypeAPIHelper.createEmptySampleType(getProjectName(), props);
-        sampleSetDataGenerator.generateRows(30);
-        for (int i=0; i < 5; i++)  // add some rows with a description we can filter on
+        sampleSetDataGenerator.generateRows(30, List.of("descColumn"));
```
Consider setting a custom data generator for 'descColumn' instead of overloading `generateRows`.
```suggestion
        sampleSetDataGenerator.addDataSupplier("descColumn", () -> null);
        sampleSetDataGenerator.generateRows(30);
```

cnathe replied:
Perfect. I wasn't aware of that. Changes made.

cnathe committed:
CR feedback
- clearUserDefinedFilters() -> clearFilters()
- remove FilterStatus component and move getters
- remove sortOn()
- add confirmExpectingError()

cnathe committed:
Merge branch 'develop' into fb_gridFilterUXEnabled
